### PR TITLE
Ensemble/895/mithril client snapshot commands

### LINKED ISSUE #895: Create the sub-command for `Cardano Immutable Files Full` in client
## Issue
Create the sub-command for `Cardano Immutable Files Full`:
- `mithril-client snapshot list`
- `mithril-client snapshot show $SNAPSHOT_ID`
- `mithril-client snapshot download  $SNAPSHOT_ID`

## To do

- [x] Adapt previous commands to new interface with a `Snapshot` sub-command
- [x] #512

## Note
Existing commands:
- `mithril-client list`
- `mithril-client show $SNAPSHOT_ID`
- `mithril-client download  $SNAPSHOT_ID`
- `mithril-client restore  $SNAPSHOT_ID`

## Parent issues
#894 and #780

### PR DESCRIPTION
**Pull request by ghubertpalo**
## Content
This PR includes snapshot commands to the Mithril client.

## Pre-submit checklist

- Branch
  - [X] Tests are provided
  - [ ] Crates versions are updated
  - [ ] Commit sequence broadly makes sense
  - [X] Key commits have useful messages
- PR
  - [ ] No clippy warnings in the CI
  - [X] Self-reviewed the diff
  - [ ] Useful pull request description
  - [ ] Reviewer requested
- Documentation
  - [ ] Update README file
  - [ ] Update documentation website
  - [ ] Add dev blog post (yes but not in this ticket)

## Comments
This PR also includes some refactoring in the common crate.

## Issue(s)
Closes #895

### CODE REVIEW

File: mithril-client/src/commands/snapshot/list.rs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                &config.aggregator_endpoint,
+                APIVersionProvider::compute_all_versions_sorted()?,
+            ));
+
+            MithrilClientSnapshotService::new(
+                Arc::new(SnapshotClient::new(http_client.clone())),
+                Arc::new(CertificateClient::new(http_client)),
+                Arc::new(MithrilCertificateVerifier::new(slog_scope::logger())),
+                Arc::new(CardanoImmutableDigester::new(
+                    Path::new(""),
```
Comment (jpraynaud) [inline mithril-client/src/commands/snapshot/list.rs:47]: Maybe add a comment or to do here to explain with the path is empty?

File: mithril-client/src/commands/snapshot/show.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                &config.aggregator_endpoint,
+                APIVersionProvider::compute_all_versions_sorted()?,
+            ));
+
+            MithrilClientSnapshotService::new(
+                Arc::new(SnapshotClient::new(http_client.clone())),
+                Arc::new(CertificateClient::new(http_client)),
+                Arc::new(MithrilCertificateVerifier::new(slog_scope::logger())),
+                Arc::new(CardanoImmutableDigester::new(
+                    Path::new(""),
```
Comment (jpraynaud) [inline mithril-client/src/commands/snapshot/show.rs:50]: Same comment as above

File: mithril-client/src/commands/snapshot/show.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+            print_stdout(convert_to_field_items(&snapshot).with_title()).unwrap();
+        }
+
+        Ok(())
+    }
+}
+
+/// Convert Snapshot to SnapshotFieldItems routine
+fn convert_to_field_items(snapshot: &Snapshot) -> Vec<SnapshotFieldItem> {
+    let mut field_items = vec![
```
Comment (jpraynaud) [inline mithril-client/src/commands/snapshot/show.rs:70]: We could add a `SnapshotFieldItem` with the Cardano network and use the `snapshot.beacon.network` to retrieve the value

File: mithril-client/src/entities.rs [hunk 4]
```diff
@@ -21,10 +21,6 @@ pub struct Config {
 /// for the purpose of tabular display
 #[derive(Table, Debug, Clone, PartialEq, Eq, PartialOrd, Serialize)]
 pub struct SnapshotListItem {
-    /// Cardano network
```
Comment (jpraynaud) [inline mithril-client/src/entities.rs:24]: We could have kept this information of the Cardano network and use the `snapshot.beacon.network` to retrieve the value

File: mithril-client/src/main.rs [hunk 5]
```diff
@@ -98,27 +93,39 @@ impl Source for Args {
...
+    ) -> Result<(), StdError> {
+        match self {
+            Self::Snapshot(cmd) => cmd.execute(config_builder).await,
+        }
+    }
+}
 /// CLI command list
 #[derive(Subcommand, Debug, Clone)]
-enum Commands {
+enum SnapshotCommands {
```
Comment (jpraynaud) [inline mithril-client/src/main.rs:114]: This could be declared in the snapshot command sub-module? WDYT?

**Review by jpraynaud — APPROVED** [review]
LGTM :+1: 

I left few comments below

**Review by jpraynaud — COMMENTED** [review]
We miss an update of this doc as well: https://mithril.network/doc/manual/getting-started/run-mithril-devnet

File: mithril-client/src/aggregator_client/certificate_client.rs [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+        Self { http_client }
+    }
+
+    /// Get a single certificate full information from the aggregator.
+    pub async fn get(&self, certificate_hash: &str) -> StdResult<Option<Certificate>> {
+        let url = format!("certificate/{certificate_hash}");
+        let response = self.http_client.get_content(&url).await;
+
+        match response {
+            Err(e) if matches!(e, AggregatorHTTPClientError::RemoteServerLogical(_)) => Ok(None),
```
Comment (Alenar) [inline mithril-client/src/aggregator_client/certificate_client.rs:33]: A test that assert that a None is returned in this case would be nice

**Review by Alenar — APPROVED** [review]

**Pull request merged**